test/e2e/storage/csi_volumes.go

```diff
@@ -143,6 +145,116 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
 		})
 	}
 
+	Context("CSI Topology test using GCE PD driver [Feature: CSINodeInfo]", func() {
+		driver := drivers.InitGcePDCSIDriver().(drivers.DynamicPVTestDriver) // HACK (verult) eliminate by moving this test to common test suite.
+		BeforeEach(func() {
+			drivers.SetCommonDriverParameters(driver, f, config)
+			driver.SkipUnsupportedTest(testpatterns.TestPattern{}) // HACK gets GCE PD driver to install secrets.
+			driver.CreateDriver()
+		})
+
+		AfterEach(func() {
+			driver.CleanupDriver()
+		})
+
+		It("should provision zonal PD with immediate volume binding and mount the volume to a pod", func() {
+			framework.SkipIfMultizone(cs)
+
+			t := createGCEPDStorageClassTest()
+			claim := newClaim(t, ns.GetName(), "")
+			class := newStorageClass(t, ns.GetName(), "")
+			claim.Spec.StorageClassName = &class.ObjectMeta.Name
+			testsuites.TestDynamicProvisioning(t, cs, claim, class)
+		})
+
+		It("should provision zonal PD with immediate volume binding and AllowedTopologies set and mount the volume to a pod", func() {
```

Comment:
It implicitly verifies by ensuring both positive and negative tests pass. I wanted to avoid inspecting topology details and instead focus on the end user's experience, which is whether or not a volume is available under different topology constraints.